Documentation for manual instrumentation (#478)
open-telemetry/opentelemetry-dotnet-instrumentation: docs/manual-instrumentation.md

manual instrumentation. By using both automatic and manual instrumentation, you can
better instrument the logic and functionality of your applications, clients, and frameworks.

To create your custom traces manually, follow these steps:

1. Add the `System.Diagnostics.DiagnosticSource` dependency to your project:

    ```xml
    <PackageReference Include="System.Diagnostics.DiagnosticSource" Version="6.0.0" />
    ```

2. Create an `ActivitySource` instance:

    ```csharp
        private static readonly ActivitySource RegisteredActivity = new ActivitySource("Examples.ManualInstrumentations.Registered");
    ```

3. Create an `Activity`. Optionally, set tags:

    ```csharp
            using (var activity = RegisteredActivity.StartActivity("Main"))
            {
                activity?.SetTag("foo", "bar1");
                // your logic for Main activity
            }
    ```

4. Register your `ActivitySource` in OpenTelemetry.AutoInstrumentation
by setting the `OTEL_DOTNET_AUTO_ADDITIONAL_SOURCES` environmental variable.
You can set the value to either `Examples.ManualInstrumentations.Registerd`
or to `Examples.ManualInstrumentations.*`, which registers the entire prefix.

> You can see a sample console application with auto (`HttpClient`) and manual

Reviewer: Since another PR is changing `auto` to `automatic`, I suggest doing the same here

```suggestion
You can see a sample console application with automatic (`HttpClient`) and manual
```

> instrumentation [here](..\examples\ManualInstrumenation\Examples.ManualInstrumentations).

Reviewer: For a markdown file, I'd suggest a `/` separator

> instrumentation [here](..\examples\ManualInstrumenation\Examples.ManualInstrumentations).

Author: It was working on Windows. Change to `/` in 908f7216fffb2af7256f6481883a3ebcca7b2b54

>  Note that an `Activity` created for `NonRegistered.ManualInstrumentations`
`ActivitySoruce` is not handled by the OpenTelemetry Automatic Instrumentation.

Further reading:

> - [OpenTelemetry.io documentation for .NET Manual Instrumentation](https://opentelemetry.io/docs/instrumentation/net/manual/)

Author: Fixed in 109512620c467e91e7873ca2f965f10fac0d0a5d

# Manually instrument a .NET application

> The auto-instrumentation provides a base you can build on by adding your own
> custom instrumentation. By using both instrumentation approaches, you'll be
> able to present a more detailed representation of the logic and functionality
> of your application, clients, and framework.

Reviewer 2: ```suggestion
The automatic instrumentation provides a base you can build on by adding your own
manual instrumentation. By using both automatic and manual instrumentation, you can
better instrument the logic and functionality of your applications, clients, and frameworks.
```

> To create your custom traces you should execute following steps.

Reviewer 2: ```suggestion
To create your custom traces manually, follow these steps:
```

1. Add the `System.Diagnostics.DiagnosticSource` dependency to your project:

    ```xml
    <PackageReference Include="System.Diagnostics.DiagnosticSource" Version="6.0.0" />
    ```

> 1. Create `ActivitySource` instance

Reviewer 2: ```suggestion
2. Create an `ActivitySource` instance:
```

Author: You can always put 1. in numbered list. It is easier to add points in the middle of the list. It will be handled correctly by md-displayer.

    ```csharp
        private static readonly ActivitySource RegisteredActivity = new ActivitySource("Examples.ManualInstrumentations.Registered");
    ```

> 1. Create `Activity` and optionally set tags

Reviewer 2: ```suggestion
3. Create an `Activity`. Optionally, set tags:
```

    ```csharp
            using (var activity = RegisteredActivity.StartActivity("Main"))
            {
                activity?.SetTag("foo", "bar1");
                // your logic for Main activity
            }
    ```

> 1. Register your `ActivitySource` in OpenTelemetry.AutoInstrumentation
> by setting enironemental variable `OTEL_DOTNET_AUTO_ADDITIONAL_SOURCES`.
> Value can be set either to very specific value `Examples.ManualInstrumentations.Registerd`
> or to `Examples.ManualInstrumentations.*` which registers whole prefix.

Reviewer 2: ```suggestion
4. Register your `ActivitySource` in OpenTelemetry.AutoInstrumentation
by setting the `OTEL_DOTNET_AUTO_ADDITIONAL_SOURCES` environmental variable.
You can set the value to either `Examples.ManualInstrumentations.Registerd`
or to `Examples.ManualInstrumentations.*`, which registers the entire prefix.
```

> You can see example console application with auto (`HttpClient`) and manual
> instrumentation [here](..\examples\ManualInstrumenation\Examples.ManualInstrumentations).

Reviewer 2: ```suggestion
You can see a sample console application with auto (`HttpClient`) and manual
instrumentation [here](..\examples\ManualInstrumenation\Examples.ManualInstrumentations).
```

> Please note that `Activity` created for `NonRegistered.ManualInstrumentations`
> `ActivitySoruce` will be not handled by the OpenTelemetry AutoInstrumentation.

Reviewer 2: ```suggestion
>  Note that an `Activity` created for `NonRegistered.ManualInstrumentations`
`ActivitySoruce` is not handled by the OpenTelemetry Automatic Instrumentation.
```
